### PR TITLE
Fixes #20487 - Add rebuild_config to hostgroups API

### PR DESCRIPTION
**Pull request by treydock**

### CODE REVIEW

File: app/controllers/api/v2/hostgroups_controller.rb [hunk 1]
```diff
@@ -94,12 +94,32 @@ def clone
...
+        results = @hostgroup.recreate_hosts_config(params[:only], params[:children_hosts])
+        failures = []
+        results.each_pair do |host, result|
+          host_failures = result.reject { |key, value| value }.keys.map{ |k| _(k) }
+          failures << "#{host}(#{host_failures.to_sentence})" unless host_failures.empty?
+        end
+        if failures.empty?
+          render_message _("Configuration successfully rebuilt."), :status => :ok
+        else
+          render_message (_("Configuration rebuild failed for: %s." % failures.to_sentence)), :status => :unprocessable_entity
```
Comment (ShimShtein) [inline app/controllers/api/v2/hostgroups_controller.rb]: You can use the standard `render_error` function, to render the error, like [here](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L156).

**Review by ShimShtein — CHANGES_REQUESTED** [review]
:+1:  works as expected, small comment though.

File: app/controllers/api/v2/hostgroups_controller.rb [hunk 2]
```diff
@@ -94,12 +94,32 @@ def clone
         process_response @hostgroup.save
       end
 
+      api :PUT, "/hostgroups/:id/rebuild_config", N_("Rebuild orchestration config")
+      param :id, :identifier, :required => true
+      param :only, Array, :desc => N_("Limit rebuild steps, valid steps are %{host_rebuild_steps}")
```
Comment (iNecas) [inline app/controllers/api/v2/hostgroups_controller.rb:109]: When does `%{host_rebuild_steps}"` get replaced by the real values?

Comment (treydock) [reply app/controllers/api/v2/hostgroups_controller.rb:109]: @iNecas I think this is set in config/initializers/apipie.rb:
 https://github.com/theforeman/foreman/blob/develop/config/initializers/apipie.rb#L25

**Comment by treydock** [comment]
@ShimShtein Used render_error in place of render_message.

**Comment by treydock** [comment]
The katello failure looks unrelated to my changes.  I'm unsure how to simplify the code to pass codeclimate.

File: app/controllers/api/v2/hostgroups_controller.rb (hunk 2 above)
Comment (iNecas) [reply app/controllers/api/v2/hostgroups_controller.rb:109]: You're right, thanks for pointers

**Comment by iNecas** [comment]
The katello test failures are being addressed in https://github.com/theforeman/foreman/pull/4826, let's wait for that to get resolved

**Comment by iNecas** [comment]
It would be also nice to cover the `recreate_hosts_config` method with some unit test to catch further regressions.

**Comment by treydock** [comment]
@iNecas Added some basic tests for `recreate_hosts_config`.

**Review by iNecas — APPROVED** [review]
Thanks, works for me. Please rebase, as there were some fixes upstream done to make the katello tests green. I'm ok with merging after that

**Comment by treydock** [comment]
@iNecas I've rebased

**Comment by iNecas** [comment]
Thanks @treydock , test failures unrelated. Merging now

**Comment by iNecas** [comment]
Also thanks to @ShimShtein

**Pull request merged**